debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java

```diff
@@ -297,34 +348,96 @@ protected abstract SchemaChangeEvent getCreateTableEvent(RelationalSnapshotConte
             throws Exception;
 
     private void createDataEvents(ChangeEventSourceContext sourceContext,
-                                  RelationalSnapshotContext<P, O> snapshotContext)
+                                  RelationalSnapshotContext<P, O> snapshotContext,
+                                  Queue<JdbcConnection> connectionPool)
             throws Exception {
-        SnapshotReceiver<P> snapshotReceiver = dispatcher.getSnapshotChangeEventReceiver();
         tryStartingSnapshot(snapshotContext);
 
-        final int tableCount = snapshotContext.capturedTables.size();
-        int tableOrder = 1;
-        LOGGER.info("Snapshotting contents of {} tables while still in transaction", tableCount);
-        for (Iterator<TableId> tableIdIterator = snapshotContext.capturedTables.iterator(); tableIdIterator.hasNext();) {
-            final TableId tableId = tableIdIterator.next();
-            snapshotContext.firstTable = tableOrder == 1;
-            snapshotContext.lastTable = !tableIdIterator.hasNext();
+        SnapshotReceiver<P> snapshotReceiver = dispatcher.getSnapshotChangeEventReceiver();
 
-            if (!sourceContext.isRunning()) {
-                throw new InterruptedException("Interrupted while snapshotting table " + tableId);
+        int snapshotMaxThreads = connectionPool.size();
+        LOGGER.info("Creating snapshot worker pool with {} worker thread(s)", snapshotMaxThreads);
+        ExecutorService executorService = Executors.newFixedThreadPool(snapshotMaxThreads);
+        CompletionService<Exception> completionService = new ExecutorCompletionService<>(executorService);
+
+        Map<TableId, String> queryTables = new HashMap<>();
+        Map<TableId, OptionalLong> rowCountTables = new LinkedHashMap<>();
+        for (TableId tableId : snapshotContext.capturedTables) {
+            final Optional<String> selectStatement = determineSnapshotSelect(snapshotContext, tableId);
+            if (selectStatement.isPresent()) {
+                LOGGER.info("For table '{}' using select statement: '{}'", tableId, selectStatement.get());
+                queryTables.put(tableId, selectStatement.get());
+
+                final OptionalLong rowCount = rowCountForTable(tableId);
+                rowCountTables.put(tableId, rowCount);
+            }
+            else {
+                LOGGER.warn("For table '{}' the select statement was not provided, skipping table", tableId);
+                snapshotProgressListener.dataCollectionSnapshotCompleted(snapshotContext.partition, tableId, 0);
             }
+        }
 
-            LOGGER.debug("Snapshotting table {}", tableId);
+        int snapshotOrderByRowCount = connectorConfig.snapshotOrderByRowCount();
+        if (snapshotOrderByRowCount != 0) {
+            LOGGER.info("Sort tables by row count {}", snapshotOrderByRowCount > 0 ? "asc" : "desc");
```

Comment:
If you really want to implement this then I'd recommend to run `SELECT COUNT(` for tables for which the faster row count was unavailable so it is enabled for all connectors.